### PR TITLE
Fix post and post transform retry

### PR DESCRIPTION
**Pull request by scw00**
Currently, the post and put request can not work together with server connection retry mechanism.
Because ATS can not replay the post content.

### CODE REVIEW

**Comment by scw00** [comment]
**test golang script **

```
package main

import (
	"net"
	"os"
	"fmt"
)

var closed int = 0
var resp string = "HTTP/1.1 200 OK\r\n" +
	"Cache-Control: max-age=3000\r\n" +
	"Date: Tue, 06 Jun 2017 12:07:49 GMT\r\n" +
	"Content-Length: 10\r\n" +
	"Content-Type: text/plain; charset=utf-8\r\n\r\n" +
	"helloworld"

func main()  {

	ls, err := net.Listen("tcp", os.Args[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}


	for {
		conn, err := ls.Accept()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		closed ++
		if closed % 2 != 0 {
			conn.Close()
			fmt.Println("times: ", closed, "closed directly")
			continue
		}
	       fmt.Println("times: ", closed)
	       go handleConnection(conn)	
	}
}

func handleConnection(con net.Conn)  {

	data := make([]byte, 40000)
	con.Read(data)

	fmt.Println(string(data))
	con.Write([]byte(resp))
	con.Close()
}
```

**output**
```
times:  1 closed directly
times:  2
POST /x HTTP/1.1
User-Agent: curl/7.35.0
Host: 127.0.0.1:12345
Accept: */*
Content-Length: 3
Content-Type: application/x-www-form-urlencoded
X-Forwarded-For: 127.0.0.1
Via: http/1.1 ubuntu-linux[5af10c97-8570-4476-8480-4db4f7abcc10] (ApacheTrafficServer/8.0.0)

// anther try
times 3
POST /x HTTP/1.1
User-Agent: curl/7.35.0
Host: 127.0.0.1:12345
Accept: */*
Content-Length: 3
Content-Type: application/x-www-form-urlencoded
X-Forwarded-For: 127.0.0.1
Via: http/1.1 ubuntu-linux[5af10c97-8570-4476-8480-4db4f7abcc10] (ApacheTrafficServer/8.0.0)

aaa
```

That we can see, we send the post request without data in previous request.

**Comment by oknet** [comment]
[approve ci]

@zwoop Is the autest system support golang script ?

**Comment by bryancall** [comment]
POST and PUT should be retryable if we haven't sent any data to the client.  The first conditional checks the method and if there have been any bytes sent.  The data should still be around if we haven't tried to write it to the origin.

Something else looks to be broken.

**Comment by zwoop** [comment]
This comes down to #947 according to @vmamidi. Should we consider backing it out?

**Comment by scw00** [comment]
@bryancall you are right. The reason is that we have already consumed the ua_buffer_reader in do_setup_post_tunnel
```
    client_request_body_bytes = post_buffer->write(ua_buffer_reader, chunked ? ua_buffer_reader->read_avail() : post_bytes);

    ua_buffer_reader->consume(client_request_body_bytes);
```
We will got nothing in the next term.  It can be fixed simply to remove this consume. And I will push another pr later and close this one.

@zwoop 
Although # 947 brought us a lot of trouble,  I think this pr has nothing to do with #947. 
For #947. IMHO, The best way to solve the problem is to revert #947 and reconsider the logic in iocore . @oknet  can do it better than the current logic .

**Comment by mingzym** [comment]
considering the trouble made after #947 we definitely should sort it out, no offense because this is the common case when we work in iocore, I am with @zwoop and @scw00

**Comment by scw00** [comment]
@oknet @zwoop @bryancall  Please review!!

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -5562,7 +5565,9 @@ HttpSM::do_setup_post_tunnel(HttpVC_t to_vc_type)
     //  header buffer into new buffer
     client_request_body_bytes = post_buffer->write(ua_buffer_reader, chunked ? ua_buffer_reader->read_avail() : post_bytes);
 
-    ua_buffer_reader->consume(client_request_body_bytes);
+    // Note: We can not consume the data sice we start retry.
+    // ua_buffer_reader will be destroy in HttpSessionClient::free
+    // ua_buffer_reader->consume(client_request_body_bytes);
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: The ua_buffer_reader only keeps the first part of request payload. If the tunnel has read the second part of the request payload, the ATS will only replay the first part of request payload when retry to another server and the second part will be lost.

**Comment by scw00** [comment]
#1994

**Comment by oknet** [comment]
[approve ci]

**Comment by oknet** [comment]
[approve ci]

**Comment by scw00** [comment]
The post request has blocked by #2123  sometimes . It will be fixed when #2123 merge.

Now, that is working well with post request.

Fake vc is running in the same thread. fvc will collect all data before it call sm . If post succeed or post failed entirely, fvc will use fvc_callcont to return state machine in tunnel_handler_fake. If we connect to server and send data successfully, the default handler (tunnel_handler_post)will be used.

If server aborted, we will setup fake vc ->server tunnel directly, and send the data collected by fvc. fvc is freed is tunnel reset or kill tunnel.

**Comment by scw00** [comment]
@oknet have another way to solve this problem and that will be better than this one. He will push a new pr later

**Pull request closed without merging**